[llvm][OpenMP] Add implicit cast to omp.atomic.read

NimishMishra opened the pull request:
Should the operands of `omp.atomic.read` differ, emit an implicit cast. In case of `struct` arguments, extract the 0-th index, emit an implicit cast if required, and store at the destination.

Fixes https://github.com/llvm/llvm-project/issues/112908

tblah commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -265,6 +265,32 @@ computeOpenMPScheduleType(ScheduleKind ClauseKind, bool HasChunks,
+  llvm::Type *XReadType = XRead->getType();
+  llvm::Type *VType = V->getType();
+  if (llvm::AllocaInst *vAlloca = dyn_cast<llvm::AllocaInst>(V))
+    VType = vAlloca->getAllocatedType();
```
Why would AllocaInst::getAllocatedType be different to the value's type?

NimishMishra replied:
Because the type of atomic pointer by default is `ptr`. We would like to know the real type (i.e. `AllocatedType`) in order to know whether we need a conversion or an extraction (in case of complex type).

tblah commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -265,6 +265,32 @@ computeOpenMPScheduleType(ScheduleKind ClauseKind, bool HasChunks,
+    // `store` will suffice.
+    return XRead;
+
+  if (XRead->getType()->isStructTy())
```
Why not use `XReadType` here?

tblah commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -265,6 +265,32 @@ computeOpenMPScheduleType(ScheduleKind ClauseKind, bool HasChunks,
+
+  if (XRead->getType()->isStructTy())
+    XRead = Builder.CreateExtractValue(XRead, /*Idxs=*/0);
+  if (VType->isIntegerTy() && XReadType->isFloatingPointTy())
```
nit: I think this can be `else if`

NimishMishra replied:
Actually if you consider the test case:

```
program atomic_array
  complex :: x(2)
  integer :: y

  x(1) = (1.0, 1.0)
  x(2) = (2.0, 2.0)
  !$omp parallel
    !$omp atomic read
      y = x(1)
  !$omp end parallel
  print *, y
end program
```

Then we need to extract the real component of `complex` into a `float`, and store it to `integer`.  We thus need this check in place to ensure the conversion takes place.

tblah reviewed (commented):
Thanks for circling back and fixing this. I have a few minor comments

Meinersbur commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -265,6 +265,32 @@ computeOpenMPScheduleType(ScheduleKind ClauseKind, bool HasChunks,
+  if (XRead->getType()->isStructTy())
+    XRead = Builder.CreateExtractValue(XRead, /*Idxs=*/0);
+  if (VType->isIntegerTy() && XReadType->isFloatingPointTy())
+    XRead = Builder.CreateFPToSI(XRead, VType);
+  else if (VType->isFloatingPointTy() && XReadType->isIntegerTy())
+    XRead = Builder.CreateSIToFP(XRead, VType);
```
An atomic read should not have implicit semantics of converting between int and fp. That could cause some trouble, e.g. rounding.

Meinersbur commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -8076,6 +8102,8 @@ OpenMPIRBuilder::createAtomicRead(const LocationDescription &Loc,
     }
   }
   checkAndEmitFlushAfterAtomic(Loc, AO, AtomicKind::Read);
+  if (XRead->getType() != V.Var->getType())
+    XRead = emitImplicitCast(Builder, XRead, V.Var);
```
The code above seems to already handle casting?

Meinersbur reviewed (commented):
I wonder we should use #101966 instead which handles type punning consistently by itself (https://github.com/llvm/llvm-project/blob/ee84c6bb3146d25f21c14d6d8e0abb794683f5ec/llvm/lib/Transforms/Utils/BuildBuiltins.cpp#L123-L131)

NimishMishra commented:
> I wonder we should use #101966 instead which handles type punning consistently by itself

I think we can. But how do go ahead with changes in Clang codegen? Will there be opposition from community if we try to merge a solution like https://github.com/llvm/llvm-project/pull/101966?

Meinersbur commented:
I would remove the changes to Clang and AtomicsExpandPass (in the first iteration). Those were for checking whether the API is sufficiently general.

Meinersbur commented:
I added `emitAtomicLoadBuiltin` and `emitAtomicStoreBuiltin` into #101966, and use it within OpenMPIRBuilder. The patch is not ready yet, tests are not updated and code duplication needs to be reduced.

However, I hope the benfits become clear:
 * Just one case instead of multiple depending on data type.
 * All data types (incl. complete) able to use the cmpxchg,load,store,atomicrmw etc instead of the libcall fallback. Only the data size matters, not the type.

NimishMishra commented:
> I added `emitAtomicLoadBuiltin` and `emitAtomicStoreBuiltin` into #101966, and use it within OpenMPIRBuilder. The patch is not ready yet, tests are not updated, tits probably not yet completely correct, and code duplication needs to be reduced.


Thanks for your work on this. I went through the patch, and it does seem this approach is better. I see this PR can also help with sized libcalls. 

I wonder whether work on this needs a separate PR, or whether we can merge with the current PR? Since, it is more generic and solves multiple issues at once. I am okay with picking this work and opening a new PR, where we can discuss improvements and iterate over it. Let me know what you think.

NimishMishra commented:
> I wonder whether work on this needs a separate PR, or whether we can merge with the current PR? Since, it is more generic and solves multiple issues at once. I am okay with picking this work and opening a new PR, where we can discuss improvements and iterate over it. Let me know what you think.

I discussed this with @kiranktp. We will work on https://github.com/llvm/llvm-project/pull/101966 as a separate PR, since this approach can help with multiple issues (like sized libcalls in atomic).

If everyone is okay, we can merge this current PR for the time being. This would unblock progress towards removal of the experimental message currently.

tblah commented:
@NimishMishra is this patch still needed?

NimishMishra committed:
[llvm][OpenMP] Add implicit cast to omp.atomic.read

NimishMishra commented:
> @NimishMishra is this patch still needed?

I'm sorry; this fell off my radar. Yes, without this, the test-cases in https://github.com/llvm/llvm-project/issues/112908 will segfault. I have addressed your comments, and have added a TODO based on what Michael suggested (and added that to our internal tracking system too). I'll work on integrating Michael's solution. In the meantime, we can probably go with this.

tblah reviewed (approved):
LGTM, thanks!
